### PR TITLE
feat: add find line segment Intersection algorithm

### PR DESCRIPTION
**Pull request by saahil-mahato**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in their comments that points to Wikipedia or another similar explanation.
- [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: Geometry/PlaneSweep.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    const intersections = []
+
+    for (const event of this.#events) {
+      const { type, segment } = event
+
+      if (type === 'start') {
+        this.#activeSet.add(segment)
+
+        // Check for intersections with neighboring active segments
+        const neighbors = Array.from(this.#activeSet).filter(
```
Comment (appgurueu) [inline Geometry/PlaneSweep.js:128]: How is this efficient in general? The active set might very well contain linearly many line segments meaning you get quadratic runtime? The crucial point of this algorithm is that active line segments are managed in a sorted set so you only have to check adjacent line segments in the sorted order.

Comment (saahil-mahato) [reply Geometry/PlaneSweep.js]: I've implemented Bently-Ottmann algorithm and improved the complexity to O((n+k) logn).

**Comment by saahil-mahato** [comment]
I have made changes. Please review and let me know if anything is missing. Thanks.

File: Geometry/FindIntersections.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
+/**
+ * @typedef {Object} Point
```
Comment (appgurueu) [inline Geometry/FindIntersections.js:2]: This, and many of the other typedefs (Segment, Event, maybe Intersection) should probably be classes; the functions on them make sense as methods (for example a find intersection method for Segment).

If you decide to keep (some of) these as "implicit" "structs" for whatever reason, the `create...` functions are still useless boilerplate which should be removed IMO.

File: Geometry/FindIntersections.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      Math.min(...b.types.map((t) => typePriority[t]))
+    )
+  })
+
+  // Process events
+  events.forEach((event) => {
+    if (event.types.includes('left')) {
+      event.segments.forEach((segment) => {
+        sweepLineStatus.push(segment)
+        sweepLineStatus.sort(compareSegmentsY)
```
Comment (appgurueu) [inline Geometry/FindIntersections.js:152]: This is O(n log n), resulting in a total time complexity of O(n² log n), which is worse than the naive brute force algorithm.

As said, the sweep line status structure needs to use a sorted set / map data structure. There should be efficient implementations of some in this repo; you need to use one.

File: Geometry/FindIntersections.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        const upper =
+          index < sweepLineStatus.length - 1 ? sweepLineStatus[index + 1] : null
+        sweepLineStatus = sweepLineStatus.filter((s) => s !== segment)
+        if (lower && upper)
+          handlePotentialIntersection(lower, upper, events, intersections)
+      })
+    }
+
+    if (event.types.includes('intersection')) {
+      // Re-check all pairs of segments at this x-coordinate for intersections
```
Comment (appgurueu) [inline Geometry/FindIntersections.js:177]: This also suffers from poor time complexity: First we do a linear filtering, then we do a quadratic brute force finding of intersections.

File: Geometry/Test/FindIntersections.test.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    })
+
+    it('should return positive if first segment is greater than second', () => {
+      const seg1 = { start: { x: 0, y: 3 }, end: { x: 1, y: 3 } }
+      const seg2 = { start: { x: -1, y: 2 }, end: { x: -2, y: 2 } }
+      expect(compareSegmentsY(seg1, seg2)).toBe(1)
+    })
+  })
+
+  describe('calculateIntersectionPoint', () => {
```
Comment (appgurueu) [inline Geometry/Test/FindIntersections.test.js:70]: Only `calculateIntersectionPoint`, `handlePotentialIntersection` and `findIntersections` really neeed tests. The rest is practically trivial and also covered transitively by sufficiently complex tests of the latter.

**Comment by appgurueu** [comment]
:confused:

**Pull request closed without merging**